### PR TITLE
fix(analytics): constrain LLM category output to candidate pairs

### PR DESCRIPTION
**Pull request by MH4GF**
# Summary

`categoryDecisionSchema` が `largeCategoryId` と `middleCategoryId` を任意の string で受けているため、「candidates の中から選ぶ」という制約が system プロンプトの文章にしか存在しません。構造化出力はその文章に縛られないので、候補に無い値を返せます。

返ってきた値は [`findCandidateById`](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/category-decision/engine.ts#L58-L71) が候補と突き合わせて弾きます。**誤ったカテゴリが書き込まれることはありません。** 壊れるものが無いので気づきにくいのですが、代わりに次のコストが発生しています。

### 1. その取引は自動分類のチャンスを失う

分類対象は「ローカル DB にまだ無い取引」に限られます。[対象を絞り込む条件](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/category-decision/engine.ts#L24-L32) に `!existingMfIds.has(transaction.mfId)` があり、その集合は [既に保存済みの mfId](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/category-decision/categorize-cash-flow.ts#L47-L54) です。

そして Save フェーズでは、[分類](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/crawler-phases.ts#L214) の直後に [保存](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/crawler-phases.ts#L239) が走ります。保存された取引は次回以降この集合に含まれるため、対象から外れます。つまり **取引ごとに自動分類のチャンスは 1 回だけ** です。その 1 回で候補外の値が返ると、その取引は未分類のまま残り、自動分類としては二度と試行されません。

### 2. その呼び出しは、別の取引に使えたはずのもの

[`llmCallsUsed` は呼び出しの前に増え](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/category-decision/engine.ts#L122-L131)、答えが使えなくても戻りません。1 回の実行で LLM を呼べる件数は `llm.maxPerRun` で制限され、[既定は 5](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/category-decision/config.ts#L10) です ([docs/setup.md](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/docs/setup.md#L354))。

未分類の取引が上限より多い場合、捨てられた 1 回の呼び出しは、**まだ一度も LLM を見ていない別の取引に使えたはずのもの** です。その取引も保存後は対象から外れます。候補外の値 1 件が、2 件の取引を落とすことになります。

### 3. 気づく手段が warn ログ 1 行しかない

出るのは [`LLM category decision ignored: ... is not in Money Forward candidates`](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/category-decision/engine.ts#L192-L196) という警告 1 行だけです。定期実行されるクローラのログを読む機会は多くないので、外から見える症状は「なぜか一部の取引が分類されない」になります。

### 根本的には、保証されていない性質に依存しています

制約がプロンプトの文章にしか無い以上、正しく動くかどうかはモデルが ID をそのまま返すかに依存します。コードはそれを要求していません。**したがってモデルを差し替えると、コードを変えていないのに失敗率が変わります。**

関連: #428

### 候補の組を 1 つの列挙にする

候補 1 件を `large:middle` という 1 つの鍵で表し、`z.enum` で列挙します。制約が「誰も守らせていない場所 (散文)」から「スキーマ」へ移ります。

これで 2 種類を防げます。

- **値そのものが壊れているもの。** 空文字や、ID に別の文字が混ざったもの
- **両方の ID は実在するが、組み合わせが存在しないもの。** 大項目 `11` (食費) と中項目 `77` (動画・音楽、本来は `13` 趣味・娯楽の配下) のような組は、独立した 2 つの string だと生成できてしまいます。1 つの選択肢に畳めば作れなくなります

返ってきた鍵は文字列分割せず、候補の `Map` から引き直して元の 2 つの ID に戻します。分割にすると、壊れた値からもっともらしい別のカテゴリを組み立てられてしまうためです。`LLMCategoryDecision` の形は変えていないので、この関数の外には影響しません。

候補が空のときは `z.enum` を構築できないため、モデルを呼ばずに `null` を返します。[唯一の呼び出し元は既に空をガードしている](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/category-decision/engine.ts#L122-L131)ので、実際の挙動は変わりません。

### スキーマ違反の現れ方が変わるので、その分も扱う

スキーマを狭めると、条件を満たさない出力の現れ方が変わります。AI SDK はクライアント側で検証して `NoObjectGeneratedError` を throw するため、そのままでは [engine の `catch`](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/category-decision/engine.ts#L122-L131) に落ち、**リクエストは成功しているのに `LLM_REQUEST_FAILED` として記録されます。**

そこで `generateText` の呼び出しをこの関数内で `try`/`catch` し、`NoObjectGeneratedError` だけを固有のコードで報告して `null` を返します。それ以外のエラーはそのまま投げ直すので、本物の通信障害はこれまでどおり `LLM_REQUEST_FAILED` として残ります。

判定は AI SDK に依存する `packages/analytics` 側で行います。`engine.ts` は `llmDecider` を注入で受ける設計で、`apps/crawler` は `ai` を依存に持たないため、この境界は維持しています。

ログにはモデルの出力を含めず、固定文言のみを出します。[「LLM失敗ログへ取引IDやprovider errorを含めない」](https://github.com/hiroppy/mf-dashboard/blob/3c3d64cff83bd8016d8638e664cc011f8041a213/apps/crawler/src/category-decision/engine.test.ts#L250-L266) が示す方針に従うためです。

報告先の `warn` は任意ではなく必須にしました。省略できるということは、条件を満たさない出力という signal を呼び出し側が黙って捨てられるということで、それはこの変更が無くそうとしている失敗の形そのものになるためです。

## この PR がやらないこと

- **LLM の判断そのものは改善しません。** 外食を食料品と判断する種類の誤りは変わりません。列挙が保証するのは「実在する候補であること」だけで、「正しい候補であること」ではありません
- **スキーマを強制しないプロバイダでは、候補外の出力そのものは防げません。** その場合は生成後の検証で落ち、`LLM_SCHEMA_MISMATCH` として記録されます。取引が未分類になる点は従来と同じです
- **`maxPerRun` による取りこぼしは対象外です。** 上限を超えた取引が一度も LLM を見ないまま保存される件は別の設計判断なので、触っていません

## Linear Issue

- N/A

### Selected Quality Characteristics

| Quality characteristic | Why it is relevant | Acceptance criteria |
| ---------------------- | ------------------ | ------------------- |
| Functional suitability | 取引ごとの自動分類の機会は 1 回きりで、候補外の値が返るとその機会を失う。判定を使えるかどうかが、そのまま分類できる件数に効く | スキーマが通す値はすべて候補へ解決でき、候補に無い値はスキーマが拒否する |
| Reliability | 制約がモデルの文章理解に依存してはいけない。また、制約を狭めたことで新たに生じる例外が、通信障害と混同されてはいけない | 制約をスキーマとして表現し、条件を満たさない出力は固有のコードで報告する。他のエラーの扱いは変えない |
| Performance efficiency | `llmCallsUsed` は呼び出し前に加算され戻らないため、使えない判定が 1 回の実行に許された呼び出し件数を消費する | 同じ呼び出し件数でより多くの取引が分類へ到達する |

### Test Techniques

| Test technique | Selection rationale | Expected coverage |
| -------------- | ------------------- | ----------------- |
| Equivalence partitioning | モデルが返す値は「実在する候補の組」「壊れた値」「形式は正しいが候補に無い組」に分かれ、扱いがそれぞれ異なる | 生成されたスキーマに対し、各クラスから検証する |
| Boundary value analysis | 候補が空だと列挙を構築できない。この境界を新しい早期リターンが扱う | 候補が空のときモデルを呼ばずに `null` を返すことを検証する |
| Decision table testing | 例外は「スキーマ違反か否か」で扱いが分かれ、片方だけを飲み込む | スキーマ違反は報告して `null`、それ以外は投げ直すことを両方向で検証する |

### Primary Test Conditions

- 候補に存在する鍵が受け入れられ、大項目と中項目へ解決されること
- 壊れた値 (`D11:42`、`:`) をスキーマが拒否すること
- 両方の ID は実在するが候補としては存在しない組 (`11:77`) をスキーマが拒否すること
- スキーマ違反で `NoObjectGeneratedError` が投げられたとき、`LLM_SCHEMA_MISMATCH` を報告して `null` を返すこと。報告にモデルの出力が含まれないこと
- スキーマ違反以外のエラーは握り潰さず、呼び出し元へそのまま伝わること
- 候補が空のとき、および LLM が無効のとき `generateText` を呼ばないこと
- 既存の保証が維持されていること。信頼できない取引データが `BEGIN_UNTRUSTED_JSON` の内側に隔離され、`accountName` と `mfId` が payload から除外されたままであること

### Out-of-Scope Risks

- `LLM_SCHEMA_MISMATCH` はログに出るだけで、取引ごとの状態としては記録しません。未分類のまま残る点は従来と同じです。これを人間の確認対象として可視化することは、この PR の範囲外です
- モデルへ渡す候補の形が `largeCategoryId` / `middleCategoryId` から単一の `categoryId` へ変わります。以前の形に合わせて調整されたプロンプトやモデルは挙動が変わりえます。列挙がモデルの返す値と同じ鍵で縛る必要があるため、意図的な変更です
- **条件を満たさない出力をモデルへ差し戻して再試行する仕組みは含みません。** 現状は 1 回投げて、満たさなければ捨てて終わりです。`maxRetries` は HTTP 層の失敗にしか効かず、`repairText` は JSON をパース可能にするためのもの (かつ `generateObject` 側) なので、どちらも候補外の値には効きません。

  再試行を入れる場合、`llmCallsUsed` が呼び出しの前に加算される以上、1 取引が `maxPerRun` の件数を複数消費します。既定の 5 では、再試行する取引が数件あるだけで、他の取引が一度も LLM を見ないまま保存されます。したがって上限を「呼び出し件数」で数えるか「取引数」で数えるかという設計判断が先に必要で、この PR の範囲を越えます
- 実際のプロバイダに対する実行は含みません。`generateText` を mock した単体テストと、SDK のモックモデルを通した確認で検証しています

## Verification

- [x] Unit tests
- [x] Type checking
- [x] Lint
- [ ] Storybook tests
- [ ] E2E tests
- [ ] Manual verification

### Commands Executed

```text
pnpm test — 7/7 packages passed (web 1075, db 406 + 32 skipped, crawler 390, analytics 197, mcp 15, date-utils 14, meta 9)
pnpm turbo typecheck — 9/9 passed
pnpm lint — no findings in the changed files
pnpm format:check — 609 files, all formatted
pnpm knip — no findings
```

実装だけ元に戻して新しいテストを残すと `categorization.test.ts` が落ちます。テストが新しい形だけでなく、報告した不具合そのものを捕まえていることの確認です。

あわせて、`generateText` を mock しない経路でも確認しました。SDK のモックモデルに候補外の値を返させたとき、例外が呼び出し元へ抜けず、`LLM_SCHEMA_MISMATCH` の報告と `null` になり、モデルの出力が報告に現れないことを確かめています。

### Checks Not Run

- Storybook tests — N/A。コンポーネントと story を変更していません
- E2E tests — N/A。変更は 1 関数の構造化出力スキーマと、その周辺のエラー処理に閉じています
- Manual verification — 実プロバイダでの実行は含みません。プロバイダ側が列挙を強制するかどうかはここでは示せていませんが、強制しない場合の経路 (生成後の検証で落ちる) は上記の確認で押さえています

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved cash-flow categorization reliability by validating AI-selected categories against available candidates.
  * Empty or invalid category selections now return no result instead of producing incorrect classifications.
  * Schema mismatches are handled gracefully with a warning, while other errors continue to surface normally.

### CODE REVIEW

File: packages/analytics/src/categorization.test.ts [hunk 1]
```diff
@@ -48,12 +50,87 @@ describe("generateCategoryDecisionWithLLM", () => {
...
+      transaction,
+      candidates,
+      warn: () => {},
+    });
 
     expect(result).toBeNull();
     expect(generateText).not.toHaveBeenCalled();
   });
 
+  test("候補が空の場合はgenerateTextを呼ばずnullを返す", async () => {
```
[the remark this answers is not in the document]
Comment (MH4GF) [reply packages/analytics/src/categorization.test.ts:63]: テンプレートの QA Plan 4 節はいずれも記載済みです。

- Selected Quality Characteristics: Functional suitability / Reliability / Performance efficiency の 3 件を、関連する理由と受け入れ基準つきで記載
- Test Techniques: Equivalence partitioning / Boundary value analysis / Decision table testing の 3 件を、選定理由と想定カバレッジつきで記載
- Primary Test Conditions: 7 件
- Out-of-Scope Risks: 4 件

本文の最終編集は 06:56 UTC、本レビューは 07:11 UTC なので、レビュー時点の本文にはこれらが含まれています。レビュー要求時点のスナップショットが使われた可能性があります。

変更は行いません。
